
# **<font color='blue'>panData</font>**
# **<font color='blue'>Data Analysis Projects with Python Language</font>**
# **<font color='blue'>Data Cleaning and Handling Techniques for Missing Values in Data Analysis</font>**
# **<font color='blue'>Part 2</font>**

# **Python Packages Used in the Project**

In [379]:
!pip install -q -U watermark

In [380]:
# 1. Imports
import math
import sys, os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import warnings
warnings.filterwarnings('ignore')

In [381]:
%reload_ext watermark
%watermark -a "panData"

Author: panData



## **Loading the Data**

https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

In [382]:
# 2. We create a list to identify missing values
missing_values_labels_list = ["n/a", "na", "undefined"]

In [383]:
# 3. Load the dataset
dataset = pd.read_csv("dataset.csv", na_values=missing_values_labels_list)

In [384]:
# 4. Shape
dataset.shape


(150001, 55)

In [385]:
# 5. Data Sample
dataset.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0              Samsung   Samsung Galaxy A5 Sm-A520F   
1   

In [386]:
# 6. Set the total number of columns to display when printing the dataframe
pd.set_option('display.max_columns', 100)

In [387]:
# 7. Data Sample
dataset.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0              Samsung   Samsung Galaxy A5 Sm-A520F   
1   

In [388]:
# 8. Loading the Data Dictionary
data_dictionary = pd.read_excel("dictionary.xlsx")

In [389]:
# 9. Shape
data_dictionary.shape

(56, 2)

In [390]:
# 10. Data Sample
data_dictionary.head(10)

Fields  \
0      bearer id   
1      Dur. (ms)   
2          Start   
3       Start ms   
4            End   
5         End ms   
6       Dur. (s)   
7           IMSI   
8  MSISDN/Number   
9           IMEI   

                                                             Description  
0                                                 xDr session identifier  
1                                      Total Duration of the xDR (in ms)  
2                          Start time of the xDR (first frame timestamp)  
3  Milliseconds offset of start time for the xDR (first frame timestamp)  
4                             End time of the xDR (last frame timestamp)  
5      Milliseconds offset of end time of the xDR (last frame timestamp)  
6                                       Total Duration of the xDR (in s)  
7                               International Mobile Subscriber Identity  
8          MS International PSTN/ISDN Number of mobile - customer number  
9                                International Mobile Equipment Identity

In [391]:
# 11. Set a large value for the maximum column width
pd.set_option('display.max_colwidth', 100)

In [392]:
# 12. Data Sample
data_dictionary.head(60)

Fields  \
0                                  bearer id   
1                                  Dur. (ms)   
2                                      Start   
3                                   Start ms   
4                                        End   
5                                     End ms   
6                                   Dur. (s)   
7                                       IMSI   
8                              MSISDN/Number   
9                                       IMEI   
10                        Last Location Name   
11                           Avg RTT DL (ms)   
12                           Avg RTT UL (ms)   
13                   Avg Bearer TP DL (kbps)   
14                   Avg Bearer TP UL (kbps)   
15               TCP DL Retrans. Vol (Bytes)   
16               TCP UL Retrans. Vol (Bytes)   
17                       DL TP < 50 Kbps (%)   
18            50 Kbps < DL TP < 250 Kbps (%)   
19             250 Kbps < DL TP < 1 Mbps (%)   
20                        DL TP > 1 Mbps (%)   
21                       UL TP < 10 Kbps (%)   
22             10 Kbps < UL TP < 50 Kbps (%)   
23            50 Kbps < UL TP < 300 Kbps (%)   
24                      UL TP > 300 Kbps (%)   
25                           HTTP DL (Bytes)   
26                           HTTP UL (Bytes)   
27                 Activity Duration DL (ms)   
28                 Activity Duration UL (ms)   
29                               Dur. (ms).1   
30                      Handset Manufacturer   
31                              Handset Type   
32           Nb of sec with 125000B < Vol DL   
33     Nb of sec with 1250B < Vol UL < 6250B   
34  Nb of sec with 31250B < Vol DL < 125000B   
35            Nb of sec with 37500B < Vol UL   
36    Nb of sec with 6250B < Vol DL < 31250B   
37    Nb of sec with 6250B < Vol UL < 37500B   
38             Nb of sec with Vol DL < 6250B   
39             Nb of sec with Vol UL < 1250B   
40                   Social Media DL (Bytes)   
41                   Social Media UL (Bytes)   
42                        YouTube DL (Bytes)   
43                        YouTube UL (Bytes)   
44                        Netflix DL (Bytes)   
45                        Netflix UL (Bytes)   
46                         Google DL (Bytes)   
47                         Google UL (Bytes)   
48                          Email DL (Bytes)   
49                          Email UL (Bytes)   
50                         Gaming DL (Bytes)   
51                         Gaming UL (Bytes)   
52                                  Other DL   
53                                  Other UL   
54                          Total DL (Bytes)   
55                          Total UL (Bytes)   

                                                                            Description  
0                                                                xDr session identifier  
1                                                     Total Duration of the xDR (in ms)  
2                                         Start time of the xDR (first frame timestamp)  
3                 Milliseconds offset of start time for the xDR (first frame timestamp)  
4                                            End time of the xDR (last frame timestamp)  
5                     Milliseconds offset of end time of the xDR (last frame timestamp)  
6                                                      Total Duration of the xDR (in s)  
7                                              International Mobile Subscriber Identity  
8                         MS International PSTN/ISDN Number of mobile - customer number  
9                                               International Mobile Equipment Identity  
10                          User location call name (2G/3G/4G) at the end of the bearer  
11                     Average Round Trip Time measurement Downlink direction (msecond)  
12                       Average Round Trip Time measurement Uplink direction (msecond)  
13                Average Bearer Throughp

## **Exploratory Analysis**

In [393]:
# 13. Info
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

In [394]:
# 14. Descriptive Statistics
dataset.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045927              1770.428647   
std               23971.878541              4625.355500   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844754                        3.069355   
std              13.038031                        6.215233   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530142   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.

It doesn’t make sense to calculate descriptive statistics for Beared Id, IMSI, MSISDN / Number, and IMEI. However, the describe() method calculates statistics for all numerical columns. These statistics are being calculated before the data is cleaned. Therefore, there may be changes after missing values and outliers are handled.

In [395]:
# 15. Shape
dataset.shape

(150001, 55)

In [396]:
# 16. Shape
data_dictionary.shape

(56, 2)

There are 150,001 rows and 55 columns in the dataframe. However, we have 56 columns with their names and descriptions in the dictionary. This means there is a column described but not included in the dataframe. Let’s identify the missing column.

In [397]:
# 17. Concatenate the dataframes
df_compare_columns = pd.concat([pd.Series(dataset.columns.tolist()), data_dictionary['Fields']],
                               axis=1)

In [398]:
df_compare_columns

0  \
0                                  Bearer Id   
1                                      Start   
2                                   Start ms   
3                                        End   
4                                     End ms   
5                                  Dur. (ms)   
6                                       IMSI   
7                              MSISDN/Number   
8                                       IMEI   
9                         Last Location Name   
10                           Avg RTT DL (ms)   
11                           Avg RTT UL (ms)   
12                   Avg Bearer TP DL (kbps)   
13                   Avg Bearer TP UL (kbps)   
14               TCP DL Retrans. Vol (Bytes)   
15               TCP UL Retrans. Vol (Bytes)   
16                       DL TP < 50 Kbps (%)   
17            50 Kbps < DL TP < 250 Kbps (%)   
18             250 Kbps < DL TP < 1 Mbps (%)   
19                        DL TP > 1 Mbps (%)   
20                       UL TP < 10 Kbps (%)   
21             10 Kbps < UL TP < 50 Kbps (%)   
22            50 Kbps < UL TP < 300 Kbps (%)   
23                      UL TP > 300 Kbps (%)   
24                           HTTP DL (Bytes)   
25                           HTTP UL (Bytes)   
26                 Activity Duration DL (ms)   
27                 Activity Duration UL (ms)   
28                               Dur. (ms).1   
29                      Handset Manufacturer   
30                              Handset Type   
31           Nb of sec with 125000B < Vol DL   
32     Nb of sec with 1250B < Vol UL < 6250B   
33  Nb of sec with 31250B < Vol DL < 125000B   
34            Nb of sec with 37500B < Vol UL   
35    Nb of sec with 6250B < Vol DL < 31250B   
36    Nb of sec with 6250B < Vol UL < 37500B   
37             Nb of sec with Vol DL < 6250B   
38             Nb of sec with Vol UL < 1250B   
39                   Social Media DL (Bytes)   
40                   Social Media UL (Bytes)   
41                         Google DL (Bytes)   
42                         Google UL (Bytes)   
43                          Email DL (Bytes)   
44                          Email UL (Bytes)   
45                        Youtube DL (Bytes)   
46                        Youtube UL (Bytes)   
47                        Netflix DL (Bytes)   
48                        Netflix UL (Bytes)   
49                         Gaming DL (Bytes)   
50                         Gaming UL (Bytes)   
51                          Other DL (Bytes)   
52                          Other UL (Bytes)   
53                          Total UL (Bytes)   
54                          Total DL (Bytes)   
55                                       NaN   

                                      Fields  
0                                  bearer id  
1                                  Dur. (ms)  
2                                      Start  
3                                   Start ms  
4                                        End  
5                                     End ms  
6                                   Dur. (s)  
7                                       IMSI  
8                              MSISDN/Number  
9                                       IMEI  
10                        Last Location Name  
11                           Avg RTT DL (ms)  
12                           Avg RTT UL (ms)  
13                   Avg Bearer TP DL (kbps)  
14                   Avg Bearer TP UL (kbps)  
15               TCP DL Retrans. Vol (Bytes)  
16               TCP UL Retrans. Vol (Bytes)  
17                       DL TP < 50 Kbps (%)  
18            50 Kbps < DL TP < 250 Kbps (%)  
19             250 Kbps < DL TP < 1 Mbps (%)  
20                        DL TP > 1 Mbps (%)  
21                       UL TP < 10 Kbps (%)  
22             10 Kbps < UL TP < 50 Kbps (%)  
23            50 Kbps < UL TP < 300 Kbps (%)  
24                      UL TP > 300 Kbps (%)  
25                           HTTP DL (Bytes)  
26                           HTTP UL 

In [399]:
# 18. Rename the columns
df_compare_columns.rename(columns={0: 'Column in Dataset', 'Fields': 'Column in Dictionary'},
                          inplace=True)

In [400]:
# 19. View
df_compare_columns

Column in Dataset  \
0                                  Bearer Id   
1                                      Start   
2                                   Start ms   
3                                        End   
4                                     End ms   
5                                  Dur. (ms)   
6                                       IMSI   
7                              MSISDN/Number   
8                                       IMEI   
9                         Last Location Name   
10                           Avg RTT DL (ms)   
11                           Avg RTT UL (ms)   
12                   Avg Bearer TP DL (kbps)   
13                   Avg Bearer TP UL (kbps)   
14               TCP DL Retrans. Vol (Bytes)   
15               TCP UL Retrans. Vol (Bytes)   
16                       DL TP < 50 Kbps (%)   
17            50 Kbps < DL TP < 250 Kbps (%)   
18             250 Kbps < DL TP < 1 Mbps (%)   
19                        DL TP > 1 Mbps (%)   
20                       UL TP < 10 Kbps (%)   
21             10 Kbps < UL TP < 50 Kbps (%)   
22            50 Kbps < UL TP < 300 Kbps (%)   
23                      UL TP > 300 Kbps (%)   
24                           HTTP DL (Bytes)   
25                           HTTP UL (Bytes)   
26                 Activity Duration DL (ms)   
27                 Activity Duration UL (ms)   
28                               Dur. (ms).1   
29                      Handset Manufacturer   
30                              Handset Type   
31           Nb of sec with 125000B < Vol DL   
32     Nb of sec with 1250B < Vol UL < 6250B   
33  Nb of sec with 31250B < Vol DL < 125000B   
34            Nb of sec with 37500B < Vol UL   
35    Nb of sec with 6250B < Vol DL < 31250B   
36    Nb of sec with 6250B < Vol UL < 37500B   
37             Nb of sec with Vol DL < 6250B   
38             Nb of sec with Vol UL < 1250B   
39                   Social Media DL (Bytes)   
40                   Social Media UL (Bytes)   
41                         Google DL (Bytes)   
42                         Google UL (Bytes)   
43                          Email DL (Bytes)   
44                          Email UL (Bytes)   
45                        Youtube DL (Bytes)   
46                        Youtube UL (Bytes)   
47                        Netflix DL (Bytes)   
48                        Netflix UL (Bytes)   
49                         Gaming DL (Bytes)   
50                         Gaming UL (Bytes)   
51                          Other DL (Bytes)   
52                          Other UL (Bytes)   
53                          Total UL (Bytes)   
54                          Total DL (Bytes)   
55                                       NaN   

                        Column in Dictionary  
0                                  bearer id  
1                                  Dur. (ms)  
2                                      Start  
3                                   Start ms  
4                                        End  
5                                     End ms  
6                                   Dur. (s)  
7                                       IMSI  
8                              MSISDN/Number  
9                                       IMEI  
10                        Last Location Name  
11                           Avg RTT DL (ms)  
12                           Avg RTT UL (ms)  
13                   Avg Bearer TP DL (kbps)  
14                   Avg Bearer TP UL (kbps)  
15               TCP DL Retrans. Vol (Bytes)  
16               TCP UL Retrans. Vol (Bytes)  
17                       DL TP < 50 Kbps (%)  
18            50 Kbps < DL TP < 250 Kbps (%)  
19             250 Kbps < DL TP < 1 Mbps (%)  
20                        DL TP > 1 Mbps (%)  
21                       UL TP < 10 Kbps (%)  
22             10 Kbps < UL TP < 50 Kbps (%)  
23            50 Kbps < UL TP < 300 Kbps (%)  
24                      UL TP > 300 Kbps (%)  
25                           HTTP DL (Bytes)  
26                   

“Dur. (Ms)” is ignored in the dataset as seen at index 1 in df_compare_columns. This is where the column order started to shift.

However, the same column name “Dur. (Ms)” appears in the dataset at index 5, while the dictionary file lists it as “Dur. (S)” at index 6. Since the measurements of both columns differ as indicated in their names, we need to verify which one is correct. To investigate this, we will use the column “Dur. (Ms) .1” found at indices 28 and 29 in both the dataset and the dictionary file.

In [401]:
# 20. Inspect the columns 'Dur. (ms)' and 'Dur. (ms).1'
dataset[['Dur. (ms)', 'Dur. (ms).1']]

Dur. (ms)   Dur. (ms).1
0       1823652.0  1.823653e+09
1       1365104.0  1.365104e+09
2       1361762.0  1.361763e+09
3       1321509.0  1.321510e+09
4       1089009.0  1.089009e+09
...           ...           ...
149996    81230.0  8.123076e+07
149997    97970.0  9.797070e+07
149998    98249.0  9.824953e+07
149999    97910.0  9.791063e+07
150000        NaN           NaN

[150001 rows x 2 columns]

It seems that the column “Dur. (Ms)” is measured in seconds. Therefore, we will rename it appropriately. We will also rename some of the columns to clarify their meaning and to match the naming style of the other columns.

In [402]:
# 21. Rename columns
dataset.rename(columns={'Dur. (ms)': 'Dur (s)',
                        'Dur. (ms).1': 'Dur (ms)',
                        'Start ms': 'Start Offset (ms)',
                        'End ms': 'End Offset (ms)'},
               inplace=True)

In [403]:
# 22. List of dataset columns
dataset.columns.tolist()

['Bearer Id',
 'Start',
 'Start Offset (ms)',
 'End',
 'End Offset (ms)',
 'Dur (s)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur (ms)',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 12

In [404]:
# 23. Shape
dataset.shape

(150001, 55)

## **24. Step 1 - Handling Missing Values**

### 1 - Identifying Missing Values
### 2 - Dropping Columns
### 3 - Backward Fill Imputation
### 4 - Forward Fill Imputation
### 5 - Imputation of Categorical Variables
### 6 - Dropping Rows

### **1.1. Identifying Missing Values**

In [405]:
# 24. Function to calculate the percentage of missing values
def calculate_missing_values_percentage(df):

    # Calculate the total number of cells in the dataset
    total_cells = np.product(df.shape)

    # Count the number of missing values per column
    missing_count = df.isnull().sum()

    # Calculate the total number of missing values
    total_missing = missing_count.sum()

    # Calculate the percentage of missing values
    print("The dataset has", round(((total_missing / total_cells) * 100), 2), "%", "of missing values.")

In [406]:
# 25. Check the percentage of missing values
calculate_missing_values_percentage(dataset)

The dataset has 12.72 % of missing values.


In [407]:
# 26. Function to calculate missing values by column
def calculate_missing_values_by_column(df):

    # Total missing values
    missing_values = df.isnull().sum()

    # Percentage of missing values
    missing_values_percent = 100 * missing_values / len(df)

    # Data type of columns with missing values
    missing_values_dtype = df.dtypes

    # Create a table with the results
    missing_values_table = pd.concat([missing_values, missing_values_percent, missing_values_dtype], axis=1)

    # Rename the columns
    missing_values_table_renamed = missing_values_table.rename(
        columns={0: 'Missing Values', 1: '% of Missing Values', 2: 'Dtype'})

    # Sort the table by percentage of missing values in descending order and remove columns with no missing values
    missing_values_table_renamed = missing_values_table_renamed[missing_values_table_renamed.iloc[:, 0] != 0] \
                                   .sort_values('% of Missing Values', ascending=False).round(2)

    # Print information
    print("The dataset has " + str(df.shape[1]) + " columns.\n"
          "Found: " + str(missing_values_table_renamed.shape[0]) + " columns with missing values.")

    if missing_values_table_renamed.shape[0] == 0:
        return

    # Return the dataframe with missing value information
    return missing_values_table_renamed

In [408]:
# 27. Create table with missing values
df_missing = calculate_missing_values_by_column(dataset)

The dataset has 55 columns.
Found: 41 columns with missing values.


In [409]:
# 28. View
df_missing

Missing Values  % of Missing Values  \
Nb of sec with 37500B < Vol UL                    130254                86.84   
Nb of sec with 6250B < Vol UL < 37500B            111843                74.56   
Nb of sec with 125000B < Vol DL                    97538                65.02   
TCP UL Retrans. Vol (Bytes)                        96649                64.43   
Nb of sec with 31250B < Vol DL < 125000B           93586                62.39   
Nb of sec with 1250B < Vol UL < 6250B              92894                61.93   
Nb of sec with 6250B < Vol DL < 31250B             88317                58.88   
TCP DL Retrans. Vol (Bytes)                        88146                58.76   
HTTP UL (Bytes)                                    81810                54.54   
HTTP DL (Bytes)                                    81474                54.32   
Avg RTT DL (ms)                                    27829                18.55   
Avg RTT UL (ms)                                    27812                18.54   
Handset Manufacturer                                9559                 6.37   
Handset Type                                        9559                 6.37   
Last Location Name                                  1153                 0.77   
MSISDN/Number                                       1066                 0.71   
Bearer Id                                            991                 0.66   
Nb of sec with Vol UL < 1250B                        793                 0.53   
10 Kbps < UL TP < 50 Kbps (%)                        792                 0.53   
UL TP > 300 Kbps (%)                                 792                 0.53   
50 Kbps < UL TP < 300 Kbps (%)                       792                 0.53   
UL TP < 10 Kbps (%)                                  792                 0.53   
Nb of sec with Vol DL < 6250B                        755                 0.50   
250 Kbps < DL TP < 1 Mbps (%)                        754                 0.50   
50 Kbps < DL TP < 250 Kbps (%)                       754                 0.50   
DL TP < 50 Kbps (%)                                  754                 0.50   
DL TP > 1 Mbps (%)                                   754                 0.50   
IMEI                                                 572                 0.38   
IMSI                                                 570                 0.38   
Start                                                  1                 0.00   
End Offset (ms)                                        1                 0.00   
Total UL (Bytes)                                       1                 0.00   
Start Offset (ms)                                      1                 0.00   
End                                                    1                 0.00   
Avg Bearer TP DL (kbps)                                1                 0.00   
Dur (s)                                                1                 0.00   
Avg Bearer TP UL (kbps)                                1                 0.00   
Dur (ms)                                               1                 0.00   
Activity Duration UL (ms)                              1                 0.00   
Activity Duration DL (ms)                              1                 0.00   
Total DL (Bytes)                                       1                 0.00   

                                            Dtype  
Nb of sec with 37500B < Vol UL            float64  
Nb of sec with 6250B < Vol UL < 37500B    float64  
Nb of sec with 125000B < Vol DL           float64  
TCP UL Retrans. Vol (Bytes)               float64  
Nb of sec with 31250B < Vol DL < 125000B  float64  
Nb of sec with 1250B < Vol UL < 6250B     float64  
Nb of sec with 6250B < Vol DL < 31250B    float64  
TCP DL Retrans. Vol (Bytes)               float64  
HTTP UL (Bytes)                           float64  
HTTP DL (Bytes)                           float64  
Avg RTT DL (ms)                           float64  
Avg RTT UL (ms)

Generally, columns with more than 50% missing values should be removed. Between 30% and 50% is optional.

But the final decision is always yours! Yes, you, Data Analyst. Just remember to always justify your decisions.

In this project, we will remove columns where missing values represent more than 30% of the variable, as we have a large number of columns with missing values and, consequently, a lot of work. We will handle variables with a low percentage of missing values and delete those with a high percentage of missing values.


### **1.2. Dropping Columns**

In [410]:
# 29. Columns to be removed
columns_to_remove = df_missing[df_missing['% of Missing Values'] >= 30.00].index.tolist()

In [411]:
# 30. Columns to be removed
columns_to_remove

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'TCP UL Retrans. Vol (Bytes)',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'TCP DL Retrans. Vol (Bytes)',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

Even though the “TCP” variables have many missing values, instead of removing them, we will apply imputation to these variables, as they may be necessary for our later analysis.

In [412]:
# 31. Columns to be removed
columns_to_remove = [col for col in columns_to_remove if col not in ['TCP UL Retrans. Vol (Bytes)',
                                                                     'TCP DL Retrans. Vol (Bytes)']]

In [413]:
# 32. Columns to be removed
columns_to_remove

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [414]:
# 33. Drop the columns and create another dataframe
cleaned_dataset = dataset.drop(columns_to_remove, axis=1)

In [415]:
# 34. Shape
cleaned_dataset.shape

(150001, 47)

Now let’s check the status of missing values in the modified dataframe.

In [416]:
# 35. Check the percentage of missing values in the cleaned dataframe
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 3.85 % of missing values.


In [417]:
# 36. Check missing values by column in the cleaned dataframe
calculate_missing_values_by_column(cleaned_dataset)

The dataset has 47 columns.
Found: 33 columns with missing values.


Missing Values  % of Missing Values    Dtype
TCP UL Retrans. Vol (Bytes)              96649                64.43  float64
TCP DL Retrans. Vol (Bytes)              88146                58.76  float64
Avg RTT DL (ms)                          27829                18.55  float64
Avg RTT UL (ms)                          27812                18.54  float64
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.3. Backward Fill Imputation**

Since the percentages of missing values for ‘TCP UL Retrans. Vol (Bytes)’ and ‘TCP DL Retrans. Vol (Bytes)’ are very high, we will apply backward fill imputation for the missing values.

In this case, using a single value like the mean or median is not advisable, as it may alter our data in an undesirable way, making most values equal to a single value.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [418]:
# 37. Backward fill imputation for missing values
# method='bfill': Bfill or backward-fill propagates the first non-null observed value backwards
# until another non-null value is encountered.
def backward_fill_missing_values(df, col):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(method='bfill')

    print(f"{count} missing values in column {col} were replaced using the backward fill method.")

In [419]:
# 38. Backward fill imputation on the 'TCP UL Retrans. Vol (Bytes)' variable
backward_fill_missing_values(cleaned_dataset, 'TCP UL Retrans. Vol (Bytes)')

96649 missing values in column TCP UL Retrans. Vol (Bytes) were replaced using the backward fill method.


In [420]:
# 39. Backward fill imputation on the 'TCP DL Retrans. Vol (Bytes)' variable
backward_fill_missing_values(cleaned_dataset, 'TCP DL Retrans. Vol (Bytes)')

88146 missing values in column TCP DL Retrans. Vol (Bytes) were replaced using the backward fill method.


### **1.4. Forward Fill Imputation**

In [421]:
# 40. Check missing values by column in the cleaned dataframe
calculate_missing_values_by_column(cleaned_dataset)

The dataset has 47 columns.
Found: 33 columns with missing values.


Missing Values  % of Missing Values    Dtype
Avg RTT DL (ms)                          27829                18.55  float64
Avg RTT UL (ms)                          27812                18.54  float64
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

The variables “Avg RTT DL (ms)” and “Avg RTT UL (ms)” have the next highest percentages of missing values, with approximately 18.5% each. Let’s check if these variables are skewed (do not follow a normal distribution) using the skew() method, which returns the skewness coefficient.

In [422]:
# 41. Check skewness for 'Avg RTT DL (ms)'
cleaned_dataset['Avg RTT DL (ms)'].skew(skipna=True)

62.90782807995961

In [423]:
# 42. Check skewness for 'Avg RTT UL (ms)'
cleaned_dataset['Avg RTT UL (ms)'].skew(skipna=True)

28.45741458546382

	•	If skewness is between -0.5 and 0.5, the data is fairly symmetrical.
	•	If skewness is between -1 and -0.5 or between 0.5 and 1, the data is moderately skewed.
	•	If skewness is less than -1 or greater than 1, the data is highly skewed.

Since both columns, “Avg RTT DL (ms)” and “Avg RTT UL (ms),” are highly positively skewed, it is advisable not to impute them with their mean. Therefore, we will use forward fill imputation.

In [424]:
# 43. Forward fill imputation for missing values
# method='ffill': Ffill or forward-fill propagates the last non-null observed value forward
# until another non-null value is encountered.
def forward_fill_missing_values(df, col):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(method='ffill')

    print(f"{count} missing values in column {col} were replaced using the forward fill method.")

In [425]:
# 44. Forward fill imputation on 'Avg RTT DL (ms)'
forward_fill_missing_values(cleaned_dataset, 'Avg RTT DL (ms)')

27829 missing values in column Avg RTT DL (ms) were replaced using the forward fill method.


In [426]:
# 45. Forward fill imputation on 'Avg RTT UL (ms)'
forward_fill_missing_values(cleaned_dataset, 'Avg RTT UL (ms)')

27812 missing values in column Avg RTT UL (ms) were replaced using the forward fill method.


We check for missing values again.

In [427]:
# 46. Check the percentage of missing values in the cleaned dataframe
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 0.44 % of missing values.


In [428]:
# 47. Check missing values by column in the cleaned dataframe
calculate_missing_values_by_column(cleaned_dataset)

The dataset has 47 columns.
Found: 31 columns with missing values.


Missing Values  % of Missing Values    Dtype
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.5. Imputation of Categorical Variables**

In [429]:
# 48. Information about the cleaned dataframe
cleaned_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Bearer Id                       149010 non-null  float64
 1   Start                           150000 non-null  object 
 2   Start Offset (ms)               150000 non-null  float64
 3   End                             150000 non-null  object 
 4   End Offset (ms)                 150000 non-null  float64
 5   Dur (s)                         150000 non-null  float64
 6   IMSI                            149431 non-null  float64
 7   MSISDN/Number                   148935 non-null  float64
 8   IMEI                            149429 non-null  float64
 9   Last Location Name              148848 non-null  object 
 10  Avg RTT DL (ms)                 150001 non-null  float64
 11  Avg RTT UL (ms)                 150001 non-null  float64
 12  Avg Bearer TP DL

Since “Handset Type” and “Handset Manufacturer” are categorical columns, it is best to impute them with the value “unknown” so as not to bias the data.

In [430]:
# 49. Fill missing values
def fill_missing_value(df, col, value):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(value)

    if isinstance(value, str):
        print(f"{count} missing values in column {col} were replaced with '{value}'.")
    else:
        print(f"{count} missing values in column {col} were replaced with {value}.")

In [431]:
# 50. Impute categorical variables
fill_missing_value(cleaned_dataset, 'Handset Type', 'unknown')
fill_missing_value(cleaned_dataset, 'Handset Manufacturer', 'unknown')

9559 missing values in column Handset Type were replaced with 'unknown'.
9559 missing values in column Handset Manufacturer were replaced with 'unknown'.


> We check for missing values again.

In [432]:
# 51. Check the percentage of missing values in the cleaned dataframe
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 0.17 % of missing values.


In [433]:
# 52. Check missing values by column in the cleaned dataframe
calculate_missing_values_by_column(cleaned_dataset)

The dataset has 47 columns.
Found: 29 columns with missing values.


Missing Values  % of Missing Values    Dtype
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.6. Dropping Rows**

Since only 0.17% of the dataset contains missing values and the total number of rows is around 150,000, dropping these rows will not have a noticeable negative impact.

In [434]:
# 53. Drop rows with missing values
def drop_rows_with_na(df):

    old = df.shape[0]

    df.dropna(inplace=True)

    new = df.shape[0]

    count = old - new

    print(f"{count} rows containing missing values were dropped.")

In [435]:
# 54. Drop rows with missing values
drop_rows_with_na(cleaned_dataset)

3114 rows containing missing values were dropped.


In [436]:
# 55. Check the percentage of missing values in the cleaned dataframe
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 0.0 % of missing values.


In [437]:
# 56. Shape
cleaned_dataset.shape

(146887, 47)

In [438]:
%watermark -a "panData"

Author: panData





---



## **Step 2 - Data Type Conversion**

In [439]:
# 57. Data types of each column in the cleaned dataframe
cleaned_dataset.dtypes

Bearer Id                         float64
Start                              object
Start Offset (ms)                 float64
End                                object
End Offset (ms)                   float64
Dur (s)                           float64
IMSI                              float64
MSISDN/Number                     float64
IMEI                              float64
Last Location Name                 object
Avg RTT DL (ms)                   float64
Avg RTT UL (ms)                   float64
Avg Bearer TP DL (kbps)           float64
Avg Bearer TP UL (kbps)           float64
TCP DL Retrans. Vol (Bytes)       float64
TCP UL Retrans. Vol (Bytes)       float64
DL TP < 50 Kbps (%)               float64
50 Kbps < DL TP < 250 Kbps (%)    float64
250 Kbps < DL TP < 1 Mbps (%)     float64
DL TP > 1 Mbps (%)                float64
UL TP < 10 Kbps (%)               float64
10 Kbps < UL TP < 50 Kbps (%)     float64
50 Kbps < UL TP < 300 Kbps (%)    float64
UL TP > 300 Kbps (%)              float64
Activity Duration DL (ms)         float64
Activity Duration UL (ms)         float64
Dur (ms)                          float64
Handset Manufacturer               object
Handset Type                       object
Nb of sec with Vol DL < 6250B     float64
Nb of sec with Vol UL < 1250B     float64
Social Media DL (Bytes)           float64
Social Media UL (Bytes)           float64
Google DL (Bytes)                 float64
Google UL (Bytes)                 float64
Email DL (Bytes)                  float64
Email UL (Bytes)                  float64
Youtube DL (Bytes)                float64
Youtube UL (Bytes)                float64
Netflix DL (Bytes)                float64
Netflix UL (Bytes)                float64
Gaming DL (Bytes)                 float64
Gaming UL (Bytes)                 float64
Other DL (Bytes)                  float64
Other UL (Bytes)                  float64
Total UL (Bytes)                  float64
Total DL (Bytes)                  float64
dtype: object

In [440]:
# 58. List of columns with data type 'object' in the cleaned dataframe
cleaned_dataset.select_dtypes(include='object').columns.tolist()

['Start', 'End', 'Last Location Name', 'Handset Manufacturer', 'Handset Type']

In [441]:
# 59. Display the first few rows of the cleaned dataframe
cleaned_dataset.head()

Bearer Id            Start  Start Offset (ms)              End  \
0  1.311448e+19   4/4/2019 12:01              770.0  4/25/2019 14:35   
1  1.311448e+19   4/9/2019 13:04              235.0   4/25/2019 8:15   
2  1.311448e+19   4/9/2019 17:42                1.0  4/25/2019 11:58   
3  1.311448e+19   4/10/2019 0:31              486.0   4/25/2019 7:36   
4  1.311448e+19  4/12/2019 20:10              565.0  4/25/2019 10:40   

   End Offset (ms)    Dur (s)          IMSI  MSISDN/Number          IMEI  \
0            662.0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1            606.0  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2            652.0  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3            171.0  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4            954.0  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A             65.0              5.0   
3                T21824A             65.0              5.0   
4                D88865A             65.0              5.0   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                      19520.0                       7230.0   
1                      19520.0                       7230.0   
2                      19520.0                       7230.0   
3                      19520.0                       7230.0   
4                      19520.0                       7230.0   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  Activity Duration DL (ms)  Activity Duration UL (ms)  \
0                   0.0                    37624.0                    38787.0   
1                   0.0                      168.0                     3560.0   
2                   0.0                        0.0                        0.0   
3                   0.0                     3330.0                    37882.0   
4                   0.0                        0.0                        0.0   

       Dur (ms) Handset Manufacturer                 Handset Type  \
0  1.823653e+09              Samsung   Samsung Galaxy A5 Sm-A520F   
1  1.365104e+09              Samsung  Samsung Galaxy J5 (Sm-J530)   
2  1.361763e+09              Samsung     Samsung Galaxy A8 (2018)   
3  1.321510e+09              unknown             

Looking at the columns, we can see that the “Start” and “End” columns are actually datetime values, even though they are labeled as objects by pandas. Besides these two columns, all other columns with object data types are actually string values. Therefore, we will convert these columns to their appropriate data types.

In [442]:
# 60. Function to convert columns to datetime
def convert_to_datetime(df, columns):
    for col in columns:
        df[col] = pd.to_datetime(df[col])

In [443]:
# 61. Convert columns to datetime
convert_to_datetime(cleaned_dataset, ['Start', 'End'])

In [444]:
# 62. Data types of each column in the cleaned dataframe
cleaned_dataset.dtypes

Bearer Id                                float64
Start                             datetime64[ns]
Start Offset (ms)                        float64
End                               datetime64[ns]
End Offset (ms)                          float64
Dur (s)                                  float64
IMSI                                     float64
MSISDN/Number                            float64
IMEI                                     float64
Last Location Name                        object
Avg RTT DL (ms)                          float64
Avg RTT UL (ms)                          float64
Avg Bearer TP DL (kbps)                  float64
Avg Bearer TP UL (kbps)                  float64
TCP DL Retrans. Vol (Bytes)              float64
TCP UL Retrans. Vol (Bytes)              float64
DL TP < 50 Kbps (%)                      float64
50 Kbps < DL TP < 250 Kbps (%)           float64
250 Kbps < DL TP < 1 Mbps (%)            float64
DL TP > 1 Mbps (%)                       float64
UL TP < 10 Kbps (%)                      float64
10 Kbps < UL TP < 50 Kbps (%)            float64
50 Kbps < UL TP < 300 Kbps (%)           float64
UL TP > 300 Kbps (%)                     float64
Activity Duration DL (ms)                float64
Activity Duration UL (ms)                float64
Dur (ms)                                 float64
Handset Manufacturer                      object
Handset Type                              object
Nb of sec with Vol DL < 6250B            float64
Nb of sec with Vol UL < 1250B            float64
Social Media DL (Bytes)                  float64
Social Media UL (Bytes)                  float64
Google DL (Bytes)                        float64
Google UL (Bytes)                        float64
Email DL (Bytes)                         float64
Email UL (Bytes)                         float64
Youtube DL (Bytes)                       float64
Youtube UL (Bytes)                       float64
Netflix DL (Bytes)                       float64
Netflix UL (Bytes)                       float64
Gaming DL (Bytes)                        float64
Gaming UL (Bytes)                        float64
Other DL (Bytes)                         float64
Other UL (Bytes)                         float64
Total UL (Bytes)                         float64
Total DL (Bytes)                         float64
dtype: object

In [445]:
# 63. Extract columns with data type 'object'
string_columns = cleaned_dataset.select_dtypes(include='object').columns.tolist()

In [446]:
# 64. View the list of columns with data type 'object'
string_columns

['Last Location Name', 'Handset Manufacturer', 'Handset Type']

In [447]:
# 65. Function to convert columns to string
def convert_to_string(df, columns):
    for col in columns:
        df[col] = df[col].astype("string")

In [448]:
# 66. Convert columns to string
convert_to_string(cleaned_dataset, string_columns)

In [449]:
# 67. Data types of each column in the cleaned dataframe
cleaned_dataset.dtypes

Bearer Id                                float64
Start                             datetime64[ns]
Start Offset (ms)                        float64
End                               datetime64[ns]
End Offset (ms)                          float64
Dur (s)                                  float64
IMSI                                     float64
MSISDN/Number                            float64
IMEI                                     float64
Last Location Name                string[python]
Avg RTT DL (ms)                          float64
Avg RTT UL (ms)                          float64
Avg Bearer TP DL (kbps)                  float64
Avg Bearer TP UL (kbps)                  float64
TCP DL Retrans. Vol (Bytes)              float64
TCP UL Retrans. Vol (Bytes)              float64
DL TP < 50 Kbps (%)                      float64
50 Kbps < DL TP < 250 Kbps (%)           float64
250 Kbps < DL TP < 1 Mbps (%)            float64
DL TP > 1 Mbps (%)                       float64
UL TP < 10 Kbps (%)                      float64
10 Kbps < UL TP < 50 Kbps (%)            float64
50 Kbps < UL TP < 300 Kbps (%)           float64
UL TP > 300 Kbps (%)                     float64
Activity Duration DL (ms)                float64
Activity Duration UL (ms)                float64
Dur (ms)                                 float64
Handset Manufacturer              string[python]
Handset Type                      string[python]
Nb of sec with Vol DL < 6250B            float64
Nb of sec with Vol UL < 1250B            float64
Social Media DL (Bytes)                  float64
Social Media UL (Bytes)                  float64
Google DL (Bytes)                        float64
Google UL (Bytes)                        float64
Email DL (Bytes)                         float64
Email UL (Bytes)                         float64
Youtube DL (Bytes)                       float64
Youtube UL (Bytes)                       float64
Netflix DL (Bytes)                       float64
Netflix UL (Bytes)                       float64
Gaming DL (Bytes)                        float64
Gaming UL (Bytes)                        float64
Other DL (Bytes)                         float64
Other UL (Bytes)                         float64
Total UL (Bytes)                         float64
Total DL (Bytes)                         float64
dtype: object

We also know that Bearer Id, IMSI, MSISDN / Number, and IMEI are unique numbers used for identification. Therefore, for better readability (and to facilitate the filters used in subsequent analyses), we will change their data types from float64 to int64.

In [450]:
# 68. List of columns for conversion
int_cols = ['Bearer Id', 'IMSI', 'MSISDN/Number', 'IMEI']

In [451]:
# 69. Function to convert columns to int
def convert_to_int(df, columns):
    for col in columns:
        df[col] = df[col].astype("int64")

In [452]:
# 70. Convert columns to int
convert_to_int(cleaned_dataset, int_cols)

In [453]:
# 71. Data types of each column in the cleaned dataframe
cleaned_dataset.dtypes

Bearer Id                                  int64
Start                             datetime64[ns]
Start Offset (ms)                        float64
End                               datetime64[ns]
End Offset (ms)                          float64
Dur (s)                                  float64
IMSI                                       int64
MSISDN/Number                              int64
IMEI                                       int64
Last Location Name                string[python]
Avg RTT DL (ms)                          float64
Avg RTT UL (ms)                          float64
Avg Bearer TP DL (kbps)                  float64
Avg Bearer TP UL (kbps)                  float64
TCP DL Retrans. Vol (Bytes)              float64
TCP UL Retrans. Vol (Bytes)              float64
DL TP < 50 Kbps (%)                      float64
50 Kbps < DL TP < 250 Kbps (%)           float64
250 Kbps < DL TP < 1 Mbps (%)            float64
DL TP > 1 Mbps (%)                       float64
UL TP < 10 Kbps (%)                      float64
10 Kbps < UL TP < 50 Kbps (%)            float64
50 Kbps < UL TP < 300 Kbps (%)           float64
UL TP > 300 Kbps (%)                     float64
Activity Duration DL (ms)                float64
Activity Duration UL (ms)                float64
Dur (ms)                                 float64
Handset Manufacturer              string[python]
Handset Type                      string[python]
Nb of sec with Vol DL < 6250B            float64
Nb of sec with Vol UL < 1250B            float64
Social Media DL (Bytes)                  float64
Social Media UL (Bytes)                  float64
Google DL (Bytes)                        float64
Google UL (Bytes)                        float64
Email DL (Bytes)                         float64
Email UL (Bytes)                         float64
Youtube DL (Bytes)                       float64
Youtube UL (Bytes)                       float64
Netflix DL (Bytes)                       float64
Netflix UL (Bytes)                       float64
Gaming DL (Bytes)                        float64
Gaming UL (Bytes)                        float64
Other DL (Bytes)                         float64
Other UL (Bytes)                         float64
Total UL (Bytes)                         float64
Total DL (Bytes)                         float64
dtype: object

In [454]:
# 72. Function to drop duplicate rows
def drop_duplicates(df):
    old = df.shape[0]
    df.drop_duplicates(inplace=True)
    new = df.shape[0]
    count = old - new
    if count == 0:
        print("No duplicate rows were found.")
    else:
        print(f"{count} duplicate rows were found and removed.")

In [455]:
# 73. Check for duplicate records and remove them if found
drop_duplicates(cleaned_dataset)

No duplicate rows were found.


As we saw in the column cleaning section, we have two duration columns, one in seconds and the other in microseconds. Let’s check if the values are equal by converting the microseconds to seconds.

In [456]:
# 74. Function to convert time unit
def convert_time_unit(df, columns, factor):
    for col in columns:
        df[col] = df[col] * factor

In [457]:
# 75. Return the rows of the two columns
temp_df = cleaned_dataset[['Dur (s)', 'Dur (ms)']].copy()

In [458]:
# 76. Display the first few rows of the temporary dataframe
temp_df.head()

Dur (s)      Dur (ms)
0  1823652.0  1.823653e+09
1  1365104.0  1.365104e+09
2  1361762.0  1.361763e+09
3  1321509.0  1.321510e+09
4  1089009.0  1.089009e+09

In [459]:
# 77. Apply the function to convert microseconds to seconds
convert_time_unit(temp_df, ['Dur (ms)'], 1/1000)

In [460]:
# 78. Display the first few rows of the temporary dataframe after conversion
temp_df.head()

Dur (s)     Dur (ms)
0  1823652.0  1823652.892
1  1365104.0  1365104.371
2  1361762.0  1361762.651
3  1321509.0  1321509.685
4  1089009.0  1089009.389

In [461]:
# 79. Comparison
temp_df['Comparison_Result'] = (temp_df['Dur (s)'] == temp_df['Dur (ms)'].apply(math.floor))

In [462]:
temp_df

Dur (s)     Dur (ms)  Comparison_Result
0       1823652.0  1823652.892               True
1       1365104.0  1365104.371               True
2       1361762.0  1361762.651               True
3       1321509.0  1321509.685               True
4       1089009.0  1089009.389               True
...           ...          ...                ...
149991    61661.0    61661.729               True
149992    84940.0    84940.610               True
149993   125622.0   125622.433               True
149994   113545.0   113545.185               True
149995    59587.0    59587.792               True

[146887 rows x 3 columns]

In [463]:
# 80. Are the two columns equal?
print(all(temp_df['Comparison_Result']))

True


This proves that, when rounded, these two columns are equal. Therefore, we will keep “Dur (ms)” as it is more precise and remove “Dur (s)”.

In [464]:
# 81. Function to drop columns
def drop_columns(df, columns):
    df.drop(columns, axis=1, inplace=True)
    count = len(columns)
    if count == 1:
        print(f"{count} column was dropped.")
    else:
        print(f"{count} columns were dropped.")

In [465]:
# 82. Drop the column
drop_columns(cleaned_dataset, ['Dur (s)'])

1 column was dropped.


## **Step 3 - Outlier Treatment**

In [466]:

# 83. Display the image
# img = mpimg.imread('outliers.png')
# plt.imshow(img_dsa)
# plt.axis('off')
# plt.show()

In [467]:
# 84. Shape of the cleaned dataframe
cleaned_dataset.shape

(146887, 46)

In [468]:
# 85. Define the OutlierHandler class
class OutlierHandler:

    # Constructor that initializes with a DataFrame
    def __init__(self, df: pd.DataFrame) -> None:
        self.df = df

    # Function to count outliers in the specified columns
    def count_outliers(self, Q1, Q3, IQR, columns):

        # Define the cutoff limit for considering a value as an outlier
        cut_off = IQR * 1.5

        # Create a temporary DataFrame with boolean values indicating outliers
        temp_df = (self.df[columns] < (Q1 - cut_off)) | (self.df[columns] > (Q3 + cut_off))

        # Return the count of outliers for each column
        return [len(temp_df[temp_df[col] == True]) for col in temp_df]

    # Function to calculate skewness of the specified columns
    def calc_skew(self, columns=None):

        # If no columns are specified, use all columns of the DataFrame
        if columns is None:
            columns = self.df.columns

        # Return the skewness measure for each column
        return [self.df[col].skew() for col in columns]

    # Function to calculate the percentage of values relative to 146887 (number of rows so far)
    def percentage(self, values):
        return [str(round(((value / 146887) * 100), 2)) + '%' for value in values]

    # Function to remove outliers in the specified columns
    def remove_outliers(self, columns):
        for col in columns:

            # Calculate the Q1 and Q3 quantiles
            Q1, Q3 = self.df[col].quantile(0.25), self.df[col].quantile(0.75)

            # Calculate the interquartile range (IQR)
            IQR = Q3 - Q1

            # Define the limits for considering a value as an outlier
            cut_off = IQR * 1.5
            lower, upper = Q1 - cut_off, Q3 + cut_off

            # Remove values considered outliers
            self.df = self.df.drop(self.df[self.df[col] > upper].index)
            self.df = self.df.drop(self.df[self.df[col] < lower].index)

    # Function to replace outliers with the fence values in the specified columns
    def replace_outliers_with_fences(self, columns):
        for col in columns:

            # Calculate the Q1 and Q3 quantiles
            Q1, Q3 = self.df[col].quantile(0.25), self.df[col].quantile(0.75)

            # Calculate the interquartile range (IQR)
            IQR = Q3 - Q1

            # Define the limits for considering a value as an outlier
            cut_off = IQR * 1.5
            lower, upper = Q1 - cut_off, Q3 + cut_off

            # Replace outliers with the fence values
            self.df[col] = np.where(self.df[col] > upper, upper, self.df[col])
            self.df[col] = np.where(self.df[col] < lower, lower, self.df[col])

    # Function to obtain a statistical summary of the specified columns
    def get_overview(self, columns) -> None:

        # Calculate various statistics for the columns
        min_val = self.df[columns].min()
        Q1 = self.df[columns].quantile(0.25)
        median = self.df[columns].quantile(0.5)
        Q3 = self.df[columns].quantile(0.75)

        max_val = self.df[columns].max()
        IQR = Q3 - Q1
        skew = self.calc_skew(columns)
        outliers = self.count_outliers(Q1, Q3, IQR, columns)
        cut_off = IQR * 1.5
        lower, upper = Q1 - cut_off, Q3 + cut_off

        # Define the column names for the new DataFrame
        new_columns = ['Column Name',
                       'Min',
                       'Q1',
                       'Median',
                       'Q3',
                       'Max',
                       'IQR',
                       'Lower Fence',
                       'Upper Fence',
                       'Skew',
                       'Num_Outliers',
                       'Percent_Outliers']

        # Create a new DataFrame with the calculated statistics
        data = zip([column for column in self.df[columns]], min_val, Q1, median, Q3, max_val, IQR, lower, upper, skew, outliers, self.percentage(outliers))
        new_df = pd.DataFrame(data=data, columns=new_columns)

        # Set 'Column Name' as the index of the new DataFrame
        new_df.set_index('Column Name', inplace=True)

        # Return the new DataFrame sorted by the number of outliers
        return new_df.sort_values('Num_Outliers', ascending=False).transpose()

In [469]:
# 86. Create the outlier handler object
outlier_handler = OutlierHandler(cleaned_dataset)

In [470]:
# 87. List of columns with data type float64
float_columns = cleaned_dataset.select_dtypes('float64').columns.tolist()

In [471]:
float_columns

['Start Offset (ms)',
 'End Offset (ms)',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur (ms)',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Social Media DL (Bytes)',
 'Social Media UL (Bytes)',
 'Google DL (Bytes)',
 'Google UL (Bytes)',
 'Email DL (Bytes)',
 'Email UL (Bytes)',
 'Youtube DL (Bytes)',
 'Youtube UL (Bytes)',
 'Netflix DL (Bytes)',
 'Netflix UL (Bytes)',
 'Gaming DL (Bytes)',
 'Gaming UL (Bytes)',
 'Other DL (Bytes)',
 'Other UL (Bytes)',
 'Total UL (Bytes)',
 'Total DL (Bytes)']

In [490]:
# 88. Overview of outliers
outlier_summary = outlier_handler.get_overview(float_columns)
outlier_summary

Column Name      Start Offset (ms) Nb of sec with Vol DL < 6250B  \
Min                            0.0                           1.0   
Q1                           250.0                          87.0   
Median                       499.0                         201.0   
Q3                           749.0                        2612.5   
Max                          999.0                       6400.75   
IQR                          499.0                        2525.5   
Lower Fence                 -498.5                      -3701.25   
Upper Fence                 1497.5                       6400.75   
Skew                      0.000823                      1.205814   
Num_Outliers                     0                             0   
Percent_Outliers              0.0%                          0.0%   

Column Name      Social Media DL (Bytes) Social Media UL (Bytes)  \
Min                                 12.0                     0.0   
Q1                              898089.0                 16431.0   
Median                         1793409.0                 32908.0   
Q3                             2694640.0                 49326.5   
Max                            3586064.0                 65870.0   
IQR                            1796551.0                 32895.5   
Lower Fence                   -1796737.5               -32912.25   
Upper Fence                    5389466.5                98669.75   
Skew                           -0.001633                0.000258   
Num_Outliers                           0                       0   
Percent_Outliers                    0.0%                    0.0%   

Column Name      Google DL (Bytes) Google UL (Bytes) Email DL (Bytes)  \
Min                          207.0               3.0             14.0   
Q1                       2882586.5         1024371.5         892573.0   
Median                   5766724.0         2054793.0        1793613.0   
Q3                       8625551.0         3088071.5        2689818.0   
Max                     11462832.0         4121357.0        3586146.0   
IQR                      5742964.5         2063700.0        1797245.0   
Lower Fence            -5731860.25        -2071178.5       -1803294.5   
Upper Fence            17239997.75         6183621.5        5385685.5   
Skew                     -0.008414          0.002242        -0.002659   
Num_Outliers                     0                 0                0   
Percent_Outliers              0.0%              0.0%             0.0%   

Column Name      Email UL (Bytes) Youtube DL (Bytes) Youtube UL (Bytes)  \
Min                           2.0               53.0              105.0   
Q1                       233439.0          5833462.0          5516549.0   
Median                   466357.0         11616334.0         11013447.0   
Q3                       700297.5         17440635.5         16514278.0   
Max                      936418.0         23259098.0         22011962.0   
IQR                      466858.5         11607173.5         10997729.0   
Lower Fence            -466848.75       -11577298.25        -10980044.5   
Upper Fence            1400585.25        34851395.75         33010871.5   
Skew                     0.007312           0.000119          -0.001265   
Num_Outliers                    0                  0                  0   
Percent_Outliers             0.0%               0.0%               0.0%   

Column Name      Netflix DL (Bytes) Netflix UL (Bytes) Gaming DL (Bytes)  \
Min                            42.0               35.0            2516.0   
Q1                        5776625.5          5480202.0       210381659.0   
Median                   11642708.0         10999348.0       423492394.0   
Q3                       17469653.0         16503222.5       633333998.5   
Max                      23259189.0         22011955.0       843441889.0   
IQR                      11693027.5         11023020.5       422952339.5   
Lower Fence            -11762915.75       -11054328.75  

In [473]:
# 89. Replace outliers with the fence values
outlier_handler.replace_outliers_with_fences(float_columns)

In [474]:
# 90. Overview of outliers after replacement
outlier_summary_updated = outlier_handler.get_overview(float_columns)

## **Delivering the Results of the Analysis to Decision Makers**

Is having the sum of the upload and download data volumes for each application as a total necessary for analyses?

In [475]:
# 91. Calculate the total social media data volume
cleaned_dataset['Social Media Data Volume (Bytes)'] = cleaned_dataset['Social Media UL (Bytes)'] + cleaned_dataset['Social Media DL (Bytes)']

In [476]:
# 92. Calculate the total Google data volume
cleaned_dataset['Google Data Volume (Bytes)'] = cleaned_dataset['Google UL (Bytes)'] + cleaned_dataset['Google DL (Bytes)']

In [477]:
# 93. Calculate the total Email data volume
cleaned_dataset['Email Data Volume (Bytes)'] = cleaned_dataset['Email UL (Bytes)'] + cleaned_dataset['Email DL (Bytes)']

In [478]:
# 94. Calculate the total YouTube data volume
cleaned_dataset['Youtube Data Volume (Bytes)'] = cleaned_dataset['Youtube UL (Bytes)'] + cleaned_dataset['Youtube DL (Bytes)']

In [479]:
# 95. Calculate the total Netflix data volume
cleaned_dataset['Netflix Data Volume (Bytes)'] = cleaned_dataset['Netflix UL (Bytes)'] + cleaned_dataset['Netflix DL (Bytes)']

In [480]:
# 96. Calculate the total Gaming data volume
cleaned_dataset['Gaming Data Volume (Bytes)'] = cleaned_dataset['Gaming UL (Bytes)'] + cleaned_dataset['Gaming DL (Bytes)']

In [481]:
# 97. Calculate the total Other data volume
cleaned_dataset['Other Data Volume (Bytes)'] = cleaned_dataset['Other UL (Bytes)'] + cleaned_dataset['Other DL (Bytes)']

In [482]:
# 98. Calculate the total data volume
cleaned_dataset['Total Data Volume (Bytes)'] = cleaned_dataset['Total UL (Bytes)'] + cleaned_dataset['Total DL (Bytes)']

In [483]:
# 99. Information about the cleaned dataframe
cleaned_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 146887 entries, 0 to 149995
Data columns (total 54 columns):
 #   Column                            Non-Null Count   Dtype         
---  ------                            --------------   -----         
 0   Bearer Id                         146887 non-null  int64         
 1   Start                             146887 non-null  datetime64[ns]
 2   Start Offset (ms)                 146887 non-null  float64       
 3   End                               146887 non-null  datetime64[ns]
 4   End Offset (ms)                   146887 non-null  float64       
 5   IMSI                              146887 non-null  int64         
 6   MSISDN/Number                     146887 non-null  int64         
 7   IMEI                              146887 non-null  int64         
 8   Last Location Name                146887 non-null  string        
 9   Avg RTT DL (ms)                   146887 non-null  float64       
 10  Avg RTT UL (ms)                   146

In [484]:
# 100. Shape of the cleaned dataframe
cleaned_dataset.shape

(146887, 54)

In [485]:
# 101. Display the first few rows of the cleaned dataframe
cleaned_dataset.head()

Bearer Id               Start  Start Offset (ms)  \
0 -9223372036854775808 2019-04-04 12:01:00              770.0   
1 -9223372036854775808 2019-04-09 13:04:00              235.0   
2 -9223372036854775808 2019-04-09 17:42:00                1.0   
3 -9223372036854775808 2019-04-10 00:31:00              486.0   
4 -9223372036854775808 2019-04-12 20:10:00              565.0   

                  End  End Offset (ms)             IMSI  MSISDN/Number  \
0 2019-04-25 14:35:00            662.0  208201448079117    33664962239   
1 2019-04-25 08:15:00            606.0  208201909211140    33681854413   
2 2019-04-25 11:58:00            652.0  208200314458056    33760627129   
3 2019-04-25 07:36:00            171.0  208201402342131    33750343200   
4 2019-04-25 10:40:00            954.0  208201401415120    33699795932   

             IMEI     Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  35521209507511  9.16456699548519E+015             42.0              5.0   
1  35794009006359                L77566A             65.0              5.0   
2  35281510359387                D42335A             65.0              5.0   
3  35356610164913                T21824A             65.0              5.0   
4  35407009745539                D88865A             65.0              5.0   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                      19520.0                       7230.0   
1                      19520.0                       7230.0   
2                      19520.0                       7230.0   
3                      19520.0                       7230.0   
4                      19520.0                       7230.0   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  Activity Duration DL (ms)  Activity Duration UL (ms)  \
0                   0.0                    37624.0                    38787.0   
1                   0.0                      168.0                     3560.0   
2                   0.0                        0.0                        0.0   
3                   0.0                     3330.0                    37882.0   
4                   0.0                        0.0                        0.0   

      Dur (ms) Handset Manufacturer                 Handset Type  \
0  245108160.5              Samsung   Samsung Galaxy A5 Sm-A520F   
1  245108160.5              Samsung  Samsung Galaxy J5 (Sm-J530)   
2  245108160.5              Samsung     Samsung Galaxy A8 (2018)   
3  245108160.5    

## **Saving the Data After Cleaning**

In [486]:
# 102. Saving the cleaned data
cleaned_dataset.to_csv('cleaned_dataset.csv', index=False)

In [487]:
%watermark -a "panData"

Author: panData



In [488]:
%watermark

Last updated: 2024-11-06T06:31:25.117307+00:00

Python implementation: CPython
Python version       : 3.10.12
IPython version      : 7.34.0

Compiler    : GCC 11.4.0
OS          : Linux
Release     : 6.1.85+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 2
Architecture: 64bit



In [489]:
%watermark --iversions

pandas    : 2.2.2
sys       : 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
numpy     : 1.26.4
matplotlib: 3.8.0



# **The End**